pytermgui/ansi_interface.py

```diff
@@ -107,243 +142,263 @@ class Color:
 
         self.layer_offset = layer * 10
 
     @staticmethod
     def translate_hex(color: str) -> tuple[int, int, int]:
-        """Translate hex string of format #RRGGBB into an RGB tuple of integers"""
+        """
+        Translate hex string of format '#RGB' or '#RRGGBB' into an 'RGB' tuple of integers
+        """
+
+        if not _HEX_REGEX.match(color):
+            raise ValueError(f"{color} should be in '#RGB' or '#RRGGBB' format")
 
         if color.startswith("#"):
             color = color[1:]
 
-        rgb = []
-        for i in (0, 2, 4):
-            color_hex = color[i : i + 2]
-            rgb.append(int(color_hex, 16))
+        if len(color) == 3:
+            color = "".join(part * 2 for part in color)
 
-        return rgb[0], rgb[1], rgb[2]
+        rgb = int(color, 16)
+
+        return ((rgb & 0xFF0000) >> 16, (rgb & 0x00FF00) >> 8, (rgb & 0x0000FF))
 
     def __call__(self, text: str, color: ColorType, reset_color: bool = True) -> str:
-        """Color a piece of text using `color`.
+        """
+        Color a piece of text using `color`.
 
         The color can be one of 4 formats:
             - str colorname:   One of the predifined named colors. See the names dict.
             - int 0-256:       One of the 256 terminal colors.
             - str #RRGGBB:     CSS-style HEX string, without alpha.
             - tuple (0-256, 0-256, 0-256): Tuple of integers, representing an RGB color.
 
-        If `reset_color` is set a reset sequence is inserted at the end."""
+        If `reset_color` is set a reset sequence is inserted at the end.
+        """
 
-        # convert hex string to tuple[int, int, int]
-        if isinstance(color, str) and all(
-            char in hexdigits or char == "#" for char in color
-        ):
+        if isinstance(color, str):
             try:
                 color = self.translate_hex(color)
             except ValueError:
                 # value is not a hex number, but is string
                 pass
 
         if color in self.names:
-            new = self.names[str(color)]
-            assert isinstance(new, int)
-            color = new
+            color = self.names.get(color, color)
 
-        # rgb values
-        if isinstance(color, tuple):
+        # 24-bit/RGB values (16581375 ≈16m colors)
+        if isinstance(color, tuple) and all([isinstance(part, int) for part in color]):
             red, green, blue = color
             color_value = f"2;{red};{green};{blue}m"
-
-        # 8-bit colors (including 0-16)
-        elif isinstance(color, int) or color.isdigit():
+        # 8-bit/Xterm colors (256 colors)
+        elif int(color) in range(256):
             color_value = f"5;{color}m"
-
         else:
             raise NotImplementedError(
                 f"Not sure what to do with {color} of type {type(color)}"
             )
+        ending = "\x1b[0m" if reset_style else ""
+        return f"\x1b[{38 + self.layer_offset};{color_value}{text}{ending}"
```

Comment:
Why reimplement the `set_mode` functions here? It's more readable if we make use of the code we already have.